nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp

```diff
@@ -131,7 +131,17 @@ class BtServiceNode : public BT::ActionNodeBase
   BT::NodeStatus tick() override
   {
     if (!request_sent_) {
+      // reset the flag to send the request or not,
+      // allowing the user the option to set it in on_tick
+      should_send_request_ = true;
+
+      // user defined callback, may modify "should_send_request_".
       on_tick();
+
+      if (!should_send_request_) {
+        return BT::NodeStatus::FAILURE;
+      }
```

Comment:
Looking back at this @SteveMacenski, we now have the use case where we want to not send a request and succeed. E.g. the BT node sees that we are already in the target state, so no need to send a request and succeed.

Would is make sense to implement something like:

```
      // user defined callback, may modify "should_fail_without_sending_goal_" and "should_succeed_without_sending_goal_".
      on_tick();

      if (should_fail_without_sending_goal_) {
            return BT::NodeStatus::FAILURE;
      }
      if (should_succeed_without_sending_goal_) {
            return BT::NodeStatus::SUCCESS;
      }     
```

Reply:
I think that should be handled in a different BT node, if you read the BTs for AI and Robotics book, the common design pattern is to have a conditional node if a particular objective is complete to skip doing the actual task (e.g. a branch of the tree with a "if ball not in hand, --> then pick up ball" so that if the conditional node checking if the ball is in the hand will trigger or not trigger picking up the ball action)

If that's the case - which it also sounds like you may have (?) the return code for this doesn't matter because that BT node wouldn't be triggered if under the appropriate control flow node.